packages/teams-js/test/internal/communication.spec.ts

```diff
@@ -1,67 +1,378 @@
+      });
+
+      afterAll(() => {
+        communication.uninitializeCommunication();
+        GlobalVars.isFramelessWindow = false;
+      });
+
+      it('should throw if there is no parent window and no native interface on the current window', async () => {
+        app._initialize(undefined);
+        const initPromise = communication.initializeCommunication(undefined);
+        await expect(initPromise).rejects.toThrowError('Initialization Failed. No Parent window found.');
+      });
+
+      it('should receive valid initialize response from parent', async () => {
+        const initPromise = communication.initializeCommunication(undefined);
+        const initMessage = utils.findInitializeMessageOrThrow();
+
+        utils.respondToNativeMessage(initMessage, false, FrameContexts.content);
+
+        const initializeResponse = await initPromise;
+
+        const expectedResponse = {
+          context: FrameContexts.content,
+          clientType: undefined,
+          runtimeConfig: undefined,
+          clientSupportedSDKVersion: undefined,
+        };
+        expect(initializeResponse).toStrictEqual(expectedResponse);
+      });
+
+      it('Communication.currentWindow should be unchanged by initializeCommunication', async () => {
+        expect(communication.Communication.currentWindow).toStrictEqual(utils.mockWindow);
+
+        const initPromise = communication.initializeCommunication(undefined);
+        const initMessage = utils.findInitializeMessageOrThrow();
+
+        utils.respondToNativeMessage(initMessage, false, FrameContexts.content);
+
+        await initPromise;
+
+        expect(communication.Communication.currentWindow).toStrictEqual(utils.mockWindow);
+      });
+
+      it('should set Communication.parentWindow to undefined', async () => {
+        expect(utils.mockWindow.parent).toBeUndefined();
+
+        const initPromise = communication.initializeCommunication(undefined);
+        const initMessage = utils.findInitializeMessageOrThrow();
+
+        utils.respondToNativeMessage(initMessage, false, FrameContexts.content);
+
+        await initPromise;
+
+        expect(communication.Communication.parentWindow).toBeUndefined();
+      });
+
+      it('should set window.onNativeMessage for handling responses from the parent', async () => {
+        expect(utils.mockWindow.onNativeMessage).toBeUndefined();
+
+        const initPromise = communication.initializeCommunication(undefined);
+        const initMessage = utils.findInitializeMessageOrThrow();
+
+        utils.respondToNativeMessage(initMessage, false, FrameContexts.content);
+
+        await initPromise;
+
+        expect(communication.Communication.currentWindow.onNativeMessage).not.toBeUndefined();
+      });
+
+      it('if there is a parent window that IS NOT self, we will not send messages using onNativeMessage, will not register onNativeMessage, and Communication.parentWindow will be set to the parent of the curent window', async () => {
```

Comment:
>if there is a parent window that IS NOT self [](http://example.com/codeflow?start=10&length=44)

When... would... a parent window BE self?

Reply:
I don't know :( I will put that on my list of things to find out. I wanted to add a test for it though since we do have specific code that alters behavior based on this